src/Jobs/OaiRecordUpdateJob.php

```diff
@@ -0,0 +1,146 @@
+use SilverStripe\ORM\DataList;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\Versioned\Versioned;
+use Symbiote\QueuedJobs\Services\AbstractQueuedJob;
+use Symbiote\QueuedJobs\Services\QueuedJob;
+use Terraformers\OpenArchive\Models\OaiRecord;
+
+/**
+ * Note: Multiple actions could have been triggered on the DataObject (including archive/restore/etc) before this Job
+ * processes. That is totally fine. The OaiRecord will represent whatever the state of the DataObject is at the point
+ * in time when this Job is actually *processed* (as apposed to, the point in time when the Job was queued).
+ *
+ * @property string $className
+ * @property int $id
+ */
+class OaiRecordUpdateJob extends AbstractQueuedJob implements QueuedJob
+{
+
+    use Injectable;
+
+    public function hydrate(string $className, int $id): void
+    {
+        $this->className = $className;
+        $this->id = $id;
+    }
+
+    public function getTitle(): string
+    {
+        return 'Update OAI Record';
+    }
+
+    public function getJobType(): int
+    {
+        return QueuedJob::QUEUED;
+    }
+
+    public function process(): void
+    {
+        // Fetch our DataObject from a LIVE context. This covers us for when Versioned DataObjects have been unpublished
+        // but not archived. From the p.o.v. of our OAI feed, unpublished is still a "Deleted" state
+        /** @var DataObject $dataObject */
+        $dataObject = Versioned::withVersionedMode(function (): ?DataObject {
+            Versioned::set_stage(Versioned::LIVE);
+
+            return DataObject::get($this->className)->byID($this->id);
+        });
+
+        // The DataObject is not currently available in a LIVE context, so the corresponding OaiRecord should be marked
+        // as deleted
+        if (!$dataObject || !$dataObject->exists()) {
+            $this->markOaiRecordDeleted();
+            $this->isComplete = true;
+
+            return;
+        }
+
+        // Extension point for DataObjects to determine whether they should be able to update OaiRecords
+        if (in_array(false, $dataObject->extend('canUpdateOaiRecord'), true)) {
+            $this->markOaiRecordDeleted();
+            $this->isComplete = true;
+
+            return;
+        }
+
+        $this->updateOaiRecord($dataObject);
+        $this->isComplete = true;
+    }
+
+    protected function updateOaiRecord(DataObject $dataObject): void
+    {
+        $oaiRecord = $this->getOaiRecord();
+
+        // If there is no existing OaiRecord then we need to create one
+        if (!$oaiRecord || !$oaiRecord->exists()) {
+            $oaiRecord = OaiRecord::create();
+            $oaiRecord->RecordClass = $this->className;
+            $oaiRecord->RecordID = $this->id;
+        }
+
+        $oaiFields = $dataObject->config()->get('oai_fields');
+
+        foreach ($oaiFields as $oaiField => $dataObjectField) {
+            $oaiRecord->{$oaiField} = $dataObject->relField($dataObjectField);
+        }
```

Comment:
Can we please validate the fields specified in the `$oai_fields` config array before we map them to a `OaiRecord`? So it can helps developers to debug in the case that they supply some unsupported fields due to typo, etc. 
```suggestion
        foreach ($oaiFields as $oaiField => $dataObjectField) {
            if (!in_array($oaiField, OaiRecord::MANAGED_FIELDS, true)) {
                throw new Exception(sprintf('Unsupported OAI field provided: %s', $oaiField));
            }
            
            $oaiRecord->{$oaiField} = $dataObject->relField($dataObjectField);
        }
```

Reply:
Thanks, @MelissaWu-SS! Updated here:
https://github.com/silverstripe-terraformers/open-archive-initiative/pull/2/commits/23bdd9e67bd1915dd55bf23aa3f3f6279b9e0d79

I opted to collect all of the mismatched fields first before throwing the error, since we don't want devs to be fixing one field and then finding out there is another the next time the Job breaks.

Reply:
Sweet, the update looks good to me!👍